crates/bevy_pbr/src/render/skin.rs

```diff
@@ -0,0 +1,110 @@
+use bevy_asset::Assets;
+use bevy_ecs::prelude::*;
+use bevy_math::Mat4;
+use bevy_render::{
+    batching::NoAutomaticBatching,
+    mesh::skinning::{SkinnedMesh, SkinnedMeshInverseBindposes},
+    render_resource::{BufferUsages, BufferVec},
+    renderer::{RenderDevice, RenderQueue},
+    view::ViewVisibility,
+    Extract,
+};
+use bevy_transform::prelude::GlobalTransform;
+
+/// Maximum number of joints supported for skinned meshes.
+pub const MAX_JOINTS: usize = 256;
+
+#[derive(Component)]
+pub struct SkinIndex {
+    pub index: u32,
+}
+impl SkinIndex {
+    /// Index to be in address space based on [`SkinUniform`] size.
+    const fn new(start: usize) -> Self {
+        SkinIndex {
+            index: (start * std::mem::size_of::<Mat4>()) as u32,
+        }
+    }
+}
+
+#[derive(Resource)]
+pub struct SkinUniform {
+    pub buffer: BufferVec<Mat4>,
+}
+impl Default for SkinUniform {
+    fn default() -> Self {
+        Self {
+            buffer: BufferVec::new(BufferUsages::UNIFORM),
+        }
+    }
+}
+
+pub fn prepare_skins(
+    render_device: Res<RenderDevice>,
+    render_queue: Res<RenderQueue>,
+    mut uniform: ResMut<SkinUniform>,
+) {
+    if uniform.buffer.is_empty() {
+        return;
+    }
+
+    let len = uniform.buffer.len();
+    uniform.buffer.reserve(len, &render_device);
+    uniform.buffer.write_buffer(&render_device, &render_queue);
+}
+
+pub fn extract_skins(
+    mut commands: Commands,
+    mut previous_len: Local<usize>,
+    mut uniform: ResMut<SkinUniform>,
+    query: Extract<Query<(Entity, &ViewVisibility, &SkinnedMesh)>>,
+    inverse_bindposes: Extract<Res<Assets<SkinnedMeshInverseBindposes>>>,
+    joints: Extract<Query<&GlobalTransform>>,
+) {
+    uniform.buffer.clear();
+
+    let mut values = Vec::with_capacity(*previous_len);
+    let mut last_start = 0;
+
+    for (entity, view_visibility, skin) in &query {
+        if !view_visibility.get() {
+            continue;
+        }
+        // PERF: This can be expensive, can we move this to prepare?
+        let buffer = &mut uniform.buffer;
```

Comment:
the comment above this should be one line down but aside from that, is there a reason uniform.buffer is not accessed directly but instead rebound as a mutable reference? its not an issue, just wondering if it was a leftover shim from inlining the function and avoiding the parameter propagation

Reply:
Pretty close: it was leftover from another change. I split this PR off from #9902 which makes this mandatory.